### PR TITLE
Add plugin for browserstack

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractDesiredCapabilitiesConfigurer.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.openqa.selenium.remote.DesiredCapabilities;
+
+public abstract class AbstractDesiredCapabilitiesConfigurer implements DesiredCapabilitiesConfigurer
+{
+    @SuppressWarnings("unchecked")
+    protected void putCapabilityIntoMap(DesiredCapabilities capabilities, String mapKey, String capabilityKey,
+            Object capabilityValue)
+    {
+        Map<String, Object> sauceOptions = (Map<String, Object>) capabilities.getCapability(mapKey);
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractDesiredCapabilitiesConfigurer.java:30]: `sauceOptions` -> `nestedOptions`

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractDesiredCapabilitiesConfigurer.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.openqa.selenium.remote.DesiredCapabilities;
+
+public abstract class AbstractDesiredCapabilitiesConfigurer implements DesiredCapabilitiesConfigurer
+{
+    @SuppressWarnings("unchecked")
+    protected void putCapabilityIntoMap(DesiredCapabilities capabilities, String mapKey, String capabilityKey,
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractDesiredCapabilitiesConfigurer.java:27]: `putNestedCapability`

Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractDesiredCapabilitiesConfigurer.java:27]: `mapKey` -> `outerKey`

Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/AbstractDesiredCapabilitiesConfigurer.java:27]: `capabilityKey` -> `innerKey`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: .github/workflows/gradle.yml [hunk 3]
```diff
@@ -292,7 +292,7 @@ jobs:
             ./gradlew :vividus-tests:runStories -x testVividusInitialization \
                                                 -Pvividus.configuration.environments=system/mobile_app/browserstack/ios \
                                                 -Pvividus.configuration.suite=system/mobile_app/healthcheck \
-                                                -Pvividus.configuration.profiles=mobile_app/ios,browserstack/mobile_app \
```
Comment (ikalinin1) [inline .github/workflows/gradle.yml:295]: Noisy changes?

**Review by ikalinin1 — APPROVED** [review]

File: .github/workflows/gradle.yml (hunk 3 above)
Comment (uarlouski) [reply .github/workflows/gradle.yml:295]: browserstack/mobile_app should go first to override data in mobile_app/iOS, as we did this in saucelabs

File: vividus-plugin-browserstack/src/main/java/org/vividus/selenium/browserstack/BrowserStackCapabilitiesConfigurer.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    public void configure(DesiredCapabilities desiredCapabilities)
+    {
+        if (browserStackEnabled)
+        {
+            Optional.ofNullable(bddRunContext.getRunningStory())
+                    .map(RunningStory::getName)
+                    .ifPresent(n -> putNestedCapability(desiredCapabilities, "bstack:options", "sessionName", n));
```
Comment (valfirst) [inline vividus-plugin-browserstack/src/main/java/org/vividus/selenium/browserstack/BrowserStackCapabilitiesConfigurer.java:43]: `n` -> `name`

File: vividus-plugin-browserstack/src/main/resources/spring.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:schemaLocation="http://www.springframework.org/schema/beans https://www.springframework.org/schema/beans/spring-beans.xsd"
+       profile="browserstack">
+
+    <bean class="org.vividus.selenium.browserstack.BrowserStackCapabilitiesConfigurer" lazy-init="false">
```
Comment (valfirst) [inline vividus-plugin-browserstack/src/main/resources/spring.xml:7]: `lazy-init="false"` - why is it needed?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-browserstack/src/main/resources/spring.xml (hunk 5 above)
Comment (uarlouski) [reply vividus-plugin-browserstack/src/main/resources/spring.xml]: removed, misunderstood some concept about collection wiring

File: vividus-plugin-browserstack/src/test/java/org/vividus/selenium/browserstack/BrowserStackCapabilitiesConfigurerTests.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    void afterEach()
+    {
+        verifyNoMoreInteractions(bddRunContext, desiredCapabilities);
+    }
+
+    @Test
+    void shouldNotConfigureCapabilitiesIfBrowserStackInDisabled()
+    {
+        configurer.setBrowserStackEnabled(false);
+        configurer.configure(desiredCapabilities);
```
Comment (valfirst) [inline vividus-plugin-browserstack/src/test/java/org/vividus/selenium/browserstack/BrowserStackCapabilitiesConfigurerTests.java:54]: this test doesn't look correct: it has neither `assert` not `verify`
(I see the logic of `afterEach`, but that's wrong: single test logic is spread across multiple methods without explicit relation)

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-browserstack/src/test/java/org/vividus/selenium/browserstack/BrowserStackCapabilitiesConfigurerTests.java (hunk 6 above)
Comment (uarlouski) [reply vividus-plugin-browserstack/src/test/java/org/vividus/selenium/browserstack/BrowserStackCapabilitiesConfigurerTests.java:54]: hmm why single test logic is spread? afterEach() contains test logic that is applicable for all the tests in this class, additionally the verification in afterEach makes all verifications in this test redundant

File: vividus-plugin-browserstack/src/test/java/org/vividus/selenium/browserstack/BrowserStackCapabilitiesConfigurerTests.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    @Mock private IBddRunContext bddRunContext;
+    @Mock private DesiredCapabilities desiredCapabilities;
+    @InjectMocks private BrowserStackCapabilitiesConfigurer configurer;
+
+    @Test
+    void shouldNotConfigureCapabilitiesIfBrowserStackInDisabled()
+    {
+        configurer.setBrowserStackEnabled(false);
+        configurer.configure(desiredCapabilities);
+        verifyNoMoreInteractions(bddRunContext, desiredCapabilities);
```
Comment (valfirst) [inline vividus-plugin-browserstack/src/test/java/org/vividus/selenium/browserstack/BrowserStackCapabilitiesConfigurerTests.java:48]: `verifyNoMoreInteractions` -> `verifyNoInteractions`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-browserstack/src/test/java/org/vividus/selenium/browserstack/BrowserStackCapabilitiesConfigurerTests.java (hunk 6 above)
Comment (uarlouski) [reply vividus-plugin-browserstack/src/test/java/org/vividus/selenium/browserstack/BrowserStackCapabilitiesConfigurerTests.java:47]: https://github.com/vividus-framework/vividus/pull/1067 unit testing agreement added

**Pull request merged**